File: day5_part1_2023.rb

class RangeMap
  attr_accessor :src_start, :dest_start, :length

  def initialize(src_start, dest_start, length)
    @src_start = src_start
    @dest_start = dest_start
    @length = length
  end
end

def convert_number(number, ranges)
  ranges.each do |r|
    if number >= r.src_start && number < r.src_start + r.length
      return r.dest_start + (number - r.src_start)
    end
  end
  number
end

seeds = []
maps = []
current_ranges = []

File.open("input.txt").each do |line|
  if line.include?("map:")
    maps << current_ranges unless current_ranges.empty?
    current_ranges = []
  elsif line.start_with?("seeds:")
    seed_strs = line[7..-1].split(" ")
    seed_strs.each { |s| seeds << s.to_i }
  else
    numbers = line.split
    if numbers.length == 3
      src_start = numbers[1].to_i
      dest_start = numbers[0].to_i
      length = numbers[2].to_i
      current_ranges << RangeMap.new(src_start, dest_start, length)
    end
  end
end
maps << current_ranges

min_location = -1
seeds.each do |seed|
  location = seed
  maps.each do |m|
    location = convert_number(location, m)
  end

  min_location = location if min_location == -1 || location < min_location
end

puts min_location

File: objectivec/day5_part1_2023.m
#import <Foundation/Foundation.h>

@interface RangeMap : NSObject
@property (nonatomic) NSInteger srcStart;
@property (nonatomic) NSInteger destStart;
@property (nonatomic) NSInteger length;
@end

@implementation RangeMap
@end

NSInteger convertNumber(NSInteger number, NSArray<RangeMap *> *ranges) {
    for (RangeMap *r in ranges) {
        if (number >= r.srcStart && number < r.srcStart + r.length) {
            return r.destStart + (number - r.srcStart);
        }
    }
    return number;
}

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSString *filePath = @"input.txt";
        NSString *content = [NSString stringWithContentsOfFile:filePath encoding:NSUTF8StringEncoding error:nil];
        NSArray *lines = [content componentsSeparatedByString:@"\n"];

        NSMutableArray<NSNumber *> *seeds = [NSMutableArray array];
        NSMutableArray<NSArray<RangeMap *> *> *maps = [NSMutableArray array];
        NSMutableArray<RangeMap *> *currentRanges = [NSMutableArray array];

        for (NSString *line in lines) {
            if ([line containsString:@"map:"]) {
                if (currentRanges.count > 0) {
                    [maps addObject:[currentRanges copy]];
                    [currentRanges removeAllObjects];
                }
            } else if ([line hasPrefix:@"seeds:"]) {
                NSArray *seedStrs = [[line substringFromIndex:7] componentsSeparatedByString:@" "];
                for (NSString *s in seedStrs) {
                    [seeds addObject:@([s integerValue])];
                }
            } else {
                NSArray *numbers = [line componentsSeparatedByString:@" "];
                if (numbers.count == 3) {
                    RangeMap *range = [[RangeMap alloc] init];
                    range.destStart = [numbers[0] integerValue];
                    range.srcStart = [numbers[1] integerValue];
                    range.length = [numbers[2] integerValue];
                    [currentRanges addObject:range];
                }
            }
        }
        if (currentRanges.count > 0) {
            [maps addObject:[currentRanges copy]];
        }

        NSInteger minLocation = NSIntegerMax;
        for (NSNumber *seed in seeds) {
            NSInteger location = [seed integerValue];
            for (NSArray<RangeMap *> *m in maps) {
                location = convertNumber(location, m);
            }
            if (location < minLocation) {
                minLocation = location;
            }
        }

        NSLog(@"%ld", (long)minLocation);
    }
    return 0;
}